progression/src/main/java/me/mykindos/betterpvp/progression/progression/perks/IncreaseTreasureChanceFishingPerk.java

```diff
@@ -0,0 +1,104 @@
+package me.mykindos.betterpvp.progression.progression.perks;
+
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import lombok.extern.slf4j.Slf4j;
+import me.mykindos.betterpvp.core.config.Config;
+import me.mykindos.betterpvp.core.listener.BPvPListener;
+import me.mykindos.betterpvp.core.utilities.UtilMessage;
+import me.mykindos.betterpvp.core.utilities.model.WeighedList;
+import me.mykindos.betterpvp.progression.Progression;
+import me.mykindos.betterpvp.progression.model.ProgressionPerk;
+import me.mykindos.betterpvp.progression.model.ProgressionTree;
+import me.mykindos.betterpvp.progression.model.stats.ProgressionData;
+import me.mykindos.betterpvp.progression.tree.fishing.Fishing;
+import me.mykindos.betterpvp.progression.tree.fishing.event.PlayerCaughtFishEvent;
+import me.mykindos.betterpvp.progression.tree.fishing.loot.TreasureType;
+import me.mykindos.betterpvp.progression.tree.fishing.model.FishingLootType;
+import org.bukkit.Bukkit;
+import org.bukkit.entity.Player;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
+import org.bukkit.event.Listener;
+
+@BPvPListener
+@Singleton
+@Slf4j
+public class IncreaseTreasureChanceFishingPerk implements Listener, ProgressionPerk {
+    @Config(path = "fishing.perks.legend-chance.minLevel", defaultValue = "750")
+    @Inject
+    private int minLevel;
+
+    @Config(path = "fishing.perks.legend-chance.maxLevel", defaultValue = "1000")
+    @Inject
+    private int maxLevel;
+
+    @Config(path = "fishing.perks.legend-chance.increasePerLevel", defaultValue = "0.25")
+    @Inject
+    private double increasePerLevel;
+
+    @Inject
+    private Progression progression;
+
+    @Inject
+    private Fishing fishing;
+
+
+    @Override
+    public String getName() {
+        return "Increase Treasure Chance Fishing";
+    }
+
+    @Override
+    public Class<? extends ProgressionTree>[] acceptedTrees() {
+        return new Class[]{
+                Fishing.class
+        };
+    }
+
+    @Override
+    public boolean canUse(Player player, ProgressionData<?> data) {
+        return minLevel <= data.getLevel() &&
+                fishing.getLootTypes().getElements().stream().anyMatch(o -> (o instanceof TreasureType));
+    }
+
+    @EventHandler (priority = EventPriority.LOW)
+    public void onCatch(PlayerCaughtFishEvent event) {
+        //if we are already fishing up a treasure, don't change anything
+        if ((event.getLoot().getType() instanceof TreasureType)) return;
```

Comment:
Probably just wanna make this work on Swimmers so we don't re-roll fish catches and we only reduce their chance to catch mobs rather than hindering their chances at catching real fish

Reply:
This specific perk was designed to increase the chance of fishing up legends (I did make the assumption that the treasure type would be legends). So if the re-roll got a legend, it would keep it. Otherwise go with whatever was there first so that it does not mess with anything else.

Reply:
TreasureType is any type of item. We could probably add a convenience method in that loot implementation to check whether it's a legendary item or not

Reply:
that makes more sense, I don't know enough about how legends are handled to do that at the moment